galter_subjects_utils/contrib/lcsh/cli.py

```diff
@@ -0,0 +1,217 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023-2024 Northwestern University.
+#
+# galter-subjects-utils is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""LCSH Command line tool."""
+
+from functools import wraps
+from pathlib import Path
+
+import click
+from flask.cli import with_appcontext
+
+from galter_subjects_utils.adapter import converted_to_subjects
+from galter_subjects_utils.deltor import DeltasGenerator
+from galter_subjects_utils.keeptrace import KeepTrace
+from galter_subjects_utils.reader import get_rdm_subjects, read_csv, read_jsonl
+from galter_subjects_utils.writer import write_csv
+
+from .adapter import generate_replacements
+from .converter import LCSHRDMConverter, deprecated_lcsh
+from .downloader import LCSHDownloader
+from .scheme import LCSHScheme
+
+defaults = {
+    "downloads-dir": Path.cwd(),
+    "output-file": Path.cwd(),
+}
+
+
+@click.group()
+def lcsh():
+    """LCSH related commands."""
+
+
+def lcsh_download_options(f):
+    """Encapsulate common LCSH download options."""
+
+    @click.option(
+        "--downloads-dir", "-d",
+        type=click.Path(path_type=Path),
+        default=defaults["downloads-dir"]
+    )
+    @click.option(
+        "--no-cache",
+        default=False,
+        help="Re-download even if already downloaded.",
+        is_flag=True
+    )
+    @wraps(f)
+    def _wrapped(*args, **kwargs):
+        """Wrap f with common download options."""
+        return f(*args, **kwargs)
+
+    return _wrapped
+
+
+def to_lcsh_downloader_kwargs(parameters):
+    """To LCSHDownloader kwargs."""
+    result = {
+        "cache": not parameters["no_cache"],
+        "directory": Path.cwd() / parameters["downloads_dir"]
+    }
+    return result
+
+
+@lcsh.command("download")
+@lcsh_download_options
+def lcsh_download(**parameters):
+    """Download LCSH files."""
+    downloader_kwargs = to_lcsh_downloader_kwargs(parameters)
+    downloader = LCSHDownloader(**downloader_kwargs)
+    downloader.download()
+
+    print(f"Raw LCSH files written in {parameters['downloads_dir']}/")
+
+
+def to_lcsh_converter_kwargs(downloader):
+    """Provide LCSH converter args from `downloader`."""
+    return {
+        "topics": read_jsonl(downloader.terms_filepath)
+    }
+
+
+@lcsh.command("file")
```

Comment:
Change in passing: generate the initial vocabulary file as a csv now to save more space (reduction of ~30%).